### PR TITLE
Add take_along_axis

### PR DESCRIPTION
**Pull request by naoyam**
Equivalent to `numpy.take_along_axis` (https://numpy.org/doc/stable/reference/generated/numpy.take_along_axis.html)

Currently, it's just implemented as `TorchGatherOp` with a bool flag indicating non-indexed dimensions should have the same extents, which isn't the case with `torch.gather`. 

**Like `torch.gather`, the lookup tensor must still be a fusion input.**

### CODE REVIEW

File: test/test_gpu_gather_ops.cpp
Comment (naoyam) [inline test/test_gpu_gather_ops.cpp:1]: Renamed to test_gather_ops.cpp

File: test/test_gather_ops.cpp
Comment (naoyam) [inline test/test_gather_ops.cpp:1]: Almost the same as test_gpu_gather_ops.cpp except that it now has `is_take_along` flag to test both `torch.gather` and `numpy.take_along_axis`

File: test/test_gather_ops.cpp
Comment (naoyam) [inline test/test_gather_ops.cpp:446]: This is the only test that's completely new. Currently failing.

File: test/test_gpu_gather_ops.cpp
Comment (naoyam) [inline test/test_gpu_gather_ops.cpp:1]: I'll also rename this file to test_gather_ops.cpp before merging.

File: test/test_gpu_gather_ops.cpp
Comment (naoyam) [inline test/test_gpu_gather_ops.cpp:137]: Added `is_take_along` option so that both `torch.gather` and `numpy.take_along_axis` are tested. In the latter case, the index tensor has the same shape as the input tensor except for the indexed dimension

File: test/test_gpu_gather_ops.cpp
Comment (naoyam) [inline test/test_gpu_gather_ops.cpp:446]: This is the only test that's completely new. Currently failing.

File: csrc/ops/arith.cpp
Comment (naoyam) [inline csrc/ops/arith.cpp:113]: Moved to indexing.h/cpp

File: csrc/ops/arith.h
Comment (naoyam) [inline csrc/ops/arith.h:151]: Moved to indexing.h

File: csrc/ops/indexing.cpp
Comment (naoyam) [inline csrc/ops/indexing.cpp:24]: Just copied from arith.cpp

File: csrc/ops/indexing.cpp
Comment (naoyam) [inline csrc/ops/indexing.cpp:207]: This function is the only new addition

File: csrc/ops/indexing.h
Comment (naoyam) [inline csrc/ops/indexing.h:52]: I'm using `int64_t` here as we generally prefer `int64_t` over `int` to avoid repeated type casting

File: csrc/ops/indexing.h
Comment (naoyam) [inline csrc/ops/indexing.h:17]: Just moved from arith.h

File: csrc/root_domain_map.cpp
Comment (naoyam) [inline csrc/root_domain_map.cpp:90]: This is an important difference between `torch.gather` and `numpy.take_along_axis`

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

File: csrc/ir_internal_nodes.h
Comment (zasdfgbnm) [inline csrc/ir_internal_nodes.h:130]: Just some comment about naming: should we name the op as `TakeAlongAxis` or `TakeAlongDim`(the latter is the PyTorch name), and rename this parameter as `exact_size`? I don't like the name of `TorchGather` vs `Gather`.

File: csrc/ops/indexing.cpp
Comment (zasdfgbnm) [inline csrc/ops/indexing.cpp:253]: So for `gather`, we are setting almost every iterdomain as `IterType::GatherScatter`, but we are not doing so for `take_along_axis`? I don't think this is a problem, but I don't like this inconsistency

File: csrc/ops/indexing.cpp
Comment (zasdfgbnm) [inline csrc/ops/indexing.cpp:241]: Why? Doesn't numpy's `take_along_axis` support broadcasting from 1 to N?

File: test/test_gpu_gather_ops.cpp
Comment (zasdfgbnm) [inline test/test_gpu_gather_ops.cpp:77]: nit: It would also be good to rename `NVFuserTest` as `IndexingOpTest`.

File: test/test_gpu_gather_ops.cpp
Comment (zasdfgbnm) [inline test/test_gpu_gather_ops.cpp:453]: Should we also test the case where tv0 is also a broadcast tensor? I think this is supported by numpy.

File: csrc/ops/indexing.cpp
Comment (naoyam) [reply csrc/ops/indexing.cpp:253]: It shouldn't be necessary, and I'm planning to remove its use from `torch_gather` as well. Will create an issue to track.

Comment (naoyam) [reply csrc/ops/indexing.cpp:253]: https://github.com/NVIDIA/Fuser/issues/211

File: csrc/ops/indexing.cpp
Comment (naoyam) [reply csrc/ops/indexing.cpp:241]: OK, you're right. numpy allows broadcast inputs:

```
>>> a = np.array([[10]])
>>> ai = np.array([[0], [0]])
>>> np.take_along_axis(a, ai, 1)
array([[10],
       [10]])
```

File: csrc/ir_internal_nodes.h
Comment (zasdfgbnm) [reply csrc/ir_internal_nodes.h:130]: That is fine. Name cleanup is low priority and doesn't have to be in this PR. Could you create an issue to track?

Comment (naoyam) [reply csrc/ir_internal_nodes.h:130]: #214

Comment (naoyam) [reply csrc/ir_internal_nodes.h]: The parameter is renamed to `exact_sizes`

File: csrc/ops/indexing.cpp
Comment (naoyam) [reply csrc/ops/indexing.cpp]: I tried to support such a case here: https://github.com/NVIDIA/Fuser/pull/206/files#diff-30a0322eff3f4e5e4acc2bd48624c386db2cfe7434a6d3b8b49d0e5c2bf62143R233-R265

While I believe that should what we should do, the newly added test with a broadcast input, `TakeAlongBroadcastInput_CUDA`, segfaults. I suspect it has the same root cause as #93, so the test is currently disabled.

I'll put this RP on hold until #93 is fixed, but I don't expect much of this PR would change.

File: test/test_gpu_gather_ops.cpp
Comment (naoyam) [reply test/test_gpu_gather_ops.cpp]: Done

File: test/test_gpu_gather_ops.cpp
Comment (naoyam) [reply test/test_gpu_gather_ops.cpp:455]: As mentioned above, added a test, `TakeAlongBroadcastInput_CUDA`, but it doesn't work right now.

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
@zasdfgbnm I'm going to work on #93 first before finalizing this PR as the test with a broadcast input is failing. I don't expect much of this PR would change.

File: csrc/ops/indexing.cpp
Comment (zasdfgbnm) [inline csrc/ops/indexing.cpp:261]: If `inp_id` is broadcast and `idx_id` is concrete, then the output should be concrete, regardless of if `inp_id` has expand or not.

File: test/test_gpu_gather_ops.cpp
Comment (zasdfgbnm) [inline test/test_gpu_gather_ops.cpp:536]: nit: use `at::take_along_dim` instead of `at::gather`. I don't think this makes any difference, but it has better consistency.

**Review by zasdfgbnm — COMMENTED** [review]
I am OK with merging this PR first, then work on https://github.com/NVIDIA/Fuser/issues/93. It's up to you on how you would like to proceed. This PR looks good to me, the only blocker is https://github.com/NVIDIA/Fuser/pull/206/files#r1176701685

File: csrc/ops/indexing.cpp
Comment (naoyam) [reply csrc/ops/indexing.cpp:267]: Updated. Does this look good? [c4b0341](https://github.com/NVIDIA/Fuser/pull/206/commits/c4b0341ca6ae1f05c0ef83d7331b0dae1ec356fc)

**Review by zasdfgbnm — APPROVED** [review]
LGTM

File: test/test_gpu_gather_ops.cpp
Comment (naoyam) [reply test/test_gpu_gather_ops.cpp]: Updated

**Comment by naoyam** [comment]
!build

**Pull request merged**
